### PR TITLE
[HIP][NFC] Refactor managed var codegen

### PR DESCRIPTION
**Pull request by yxsamliu**
Refactor managed variable handling in codegen so that the transformation is done separately from registration.

This will allow the new driver to register the managed var in the linker wrapper.

### CODE REVIEW

File: clang/lib/CodeGen/CGCUDANV.cpp [hunk 1]
```diff
@@ -1160,9 +1152,8 @@ void CGNVCUDARuntime::createOffloadingEntries() {
 
 // Returns module constructor to be added.
 llvm::Function *CGNVCUDARuntime::finalizeModule() {
+  transformManagedVars();
```
Comment (Artem-B) [inline clang/lib/CodeGen/CGCUDANV.cpp:1155]: This does not look like "NFC" as we now perform the transform for the host compilation, too.

I assume we do have existing tests covering generation of the variables.

**Review by Artem-B — APPROVED** [review]
LGTM, sans the "NFC" part in the description.

File: clang/lib/CodeGen/CGCUDANV.cpp (hunk 1 above)
Comment (jhuber6) [reply clang/lib/CodeGen/CGCUDANV.cpp:1155]: I'm guessing we also don't have a test for `__managed__` in the external test suite?

Comment (yxsamliu) [reply clang/lib/CodeGen/CGCUDANV.cpp:1155]: we did the equivalent transformation previously during registration of managed var by calling replaceManagedVar directly, so it is actually NFC. We have lit test https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenCUDA/managed-var.cu that covers transformation in both device and host code.

We also have execution tests in hip-tests https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenCUDA/managed-var.cu

Comment (yxsamliu) [reply clang/lib/CodeGen/CGCUDANV.cpp:1155]: we do not have test for managed var in llvm-test-suite

**Review by jhuber6 — APPROVED** [review]

**Pull request merged**